pkg/reconciler/pubsub/pullsubscription/pullsubscription.go

```diff
@@ -0,0 +1,556 @@
+	var reconcileErr = r.reconcile(ctx, ps)
+
+	// If no error is returned, mark the observed generation.
+	// This has to be done before updateStatus is called.
+	if reconcileErr == nil {
+		ps.Status.ObservedGeneration = ps.Generation
+	}
+
+	if equality.Semantic.DeepEqual(original.Finalizers, ps.Finalizers) {
+		// If we didn't change finalizers then don't call updateFinalizers.
+
+	} else if _, updated, fErr := r.updateFinalizers(ctx, ps); fErr != nil {
+		logging.FromContext(ctx).Desugar().Warn("Failed to update PullSubscription finalizers", zap.Error(fErr))
+		r.Recorder.Eventf(ps, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update finalizers for PullSubscription %q: %v", ps.Name, fErr)
+		return fErr
+	} else if updated {
+		// There was a difference and updateFinalizers said it updated and did not return an error.
+		r.Recorder.Eventf(ps, corev1.EventTypeNormal, "Updated", "Updated PullSubscription %q finalizers", ps.Name)
+	}
+
+	if equality.Semantic.DeepEqual(original.Status, ps.Status) {
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+
+	} else if _, uErr := r.updateStatus(ctx, ps); uErr != nil {
+		logging.FromContext(ctx).Desugar().Warn("Failed to update ps status", zap.Error(uErr))
+		r.Recorder.Eventf(ps, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update status for PullSubscription %q: %v", ps.Name, uErr)
+		return uErr
+	} else if reconcileErr == nil {
+		// There was a difference and updateStatus did not return an error.
+		r.Recorder.Eventf(ps, corev1.EventTypeNormal, "Updated", "Updated PullSubscription %q", ps.Name)
+	}
+	if reconcileErr != nil {
+		r.Recorder.Event(ps, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
+	}
+
+	return reconcileErr
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, ps *v1alpha1.PullSubscription) error {
+	ctx = logging.WithLogger(ctx, r.Logger.With(zap.Any("pullsubscription", ps)))
+
+	ps.Status.InitializeConditions()
+
+	if ps.DeletionTimestamp != nil {
+		logging.FromContext(ctx).Desugar().Debug("Deleting Pub/Sub subscription")
+		if err := r.deleteSubscription(ctx, ps); err != nil {
+			ps.Status.MarkNoSubscription("SubscriptionDeleteFailed", "Failed to delete Pub/Sub subscription: %s", err.Error())
+			return err
+		}
+		ps.Status.MarkNoSubscription("SubscriptionDeleted", "Successfully deleted Pub/Sub subscription %q", ps.Status.SubscriptionID)
+		ps.Status.SubscriptionID = ""
+		removeFinalizer(ps)
+		return nil
+	}
+
+	// Sink is required.
+	sinkURI, err := r.resolveDestination(ctx, ps.Spec.Sink, ps)
+	if err != nil {
+		ps.Status.MarkNoSink("InvalidSink", err.Error())
+		return err
+	} else {
+		ps.Status.MarkSink(sinkURI)
+	}
+
+	// Transformer is optional.
+	if ps.Spec.Transformer != nil {
```

Comment:
Seems like if Transformer was originally set, then updated to be nil, the transformer URI will still be present. We should  clear the transformer URI if the Transformer field is nil.

Reply:
Done! updated UTs as well. Thanks!